circuit/src/dynamic_sets/mod.rs

```diff
@@ -0,0 +1,878 @@
+				let eddsa = Eddsa::new(
+					big_r_x[i].clone(),
+					big_r_y[i].clone(),
+					s[i].clone(),
+					pk_x[i].clone(),
+					pk_y[i].clone(),
+					res[0].clone(),
+				);
+				eddsa.synthesize(
+					&config.common,
+					&config.eddsa,
+					layouter.namespace(|| "eddsa"),
+				)?;
+			}
+		}
+
+		// filter peers' ops
+		let ops = {
+			let mut filtered_ops = Vec::new();
+			for i in 0..NUM_NEIGHBOURS {
+				let pk_i_x = pk_x[i].clone();
+				let pk_i_y = pk_y[i].clone();
+
+				let mut ops_i = ops[i].clone();
+
+				// Update the opinion array - pairs of (key, score)
+				for j in 0..NUM_NEIGHBOURS {
+					let set_pk_j_x = pk_x[j].clone();
+					let set_pk_j_y = pk_y[j].clone();
+					let op_pk_j_x = op_pk_x[i][j].clone();
+					let op_pk_j_y = op_pk_y[i][j].clone();
+
+					let equal_chip = IsEqualChipset::new(set_pk_j_x.clone(), op_pk_j_x.clone());
+					let is_same_pk_j_x = equal_chip
+						.synthesize(
+							&config.common,
+							&config.main,
+							layouter.namespace(|| "set_pk_j_x == op_pk_j_x"),
+						)
+						.is_ok();
+					let equal_chip = IsEqualChipset::new(set_pk_j_y.clone(), op_pk_j_y.clone());
+					let is_same_pk_j_y = equal_chip
+						.synthesize(
+							&config.common,
+							&config.main,
+							layouter.namespace(|| "set_pk_j_y == op_pk_j_y"),
+						)
+						.is_ok();
+					let is_diff_pk_j = !(is_same_pk_j_x && is_same_pk_j_y);
+
+					let equal_chip = IsEqualChipset::new(set_pk_j_x.clone(), default_pk_x.clone());
+					let is_default_pk_x = equal_chip
+						.synthesize(
+							&config.common,
+							&config.main,
+							layouter.namespace(|| "set_pk_j_x == default_pk_x"),
+						)
+						.is_ok();
+					let equal_chip = IsEqualChipset::new(set_pk_j_y.clone(), default_pk_y.clone());
+					let is_default_pk_y = equal_chip
+						.synthesize(
+							&config.common,
+							&config.main,
+							layouter.namespace(|| "set_pk_j_y == default_pk_y"),
+						)
+						.is_ok();
+					let is_pk_j_null = is_default_pk_x && is_default_pk_y;
+
+					let equal_chip = IsEqualChipset::new(set_pk_j_x.clone(), pk_i_x.clone());
+					let is_pk_i_x = equal_chip
+						.synthesize(
+							&config.common,
+							&config.main,
+							layouter.namespace(|| "set_pk_j_x == pk_i_x"),
+						)
+						.is_ok();
+					let equal_chip = IsEqualChipset::new(set_pk_j_y.clone(), pk_i_y.clone());
+					let is_pk_i_y = equal_chip
+						.synthesize(
+							&config.common,
+							&config.main,
+							layouter.namespace(|| "set_pk_j_y == op_pk_j_y"),
+						)
+						.is_ok();
+					let is_pk_i = is_pk_i_x && is_pk_i_y;
+
+					// Conditions for nullifying the score
+					// 1. set_pk_j != op_pk_j
+					// 2. set_pk_j == 0 (null or default)
+					// 3. set_pk_j == pk_i
+					if is_diff_pk_j || is_pk_j_null || is_pk_i {
+						ops_i[j] = zero.clone();
+					}
+
+					// Condition for correcting the pk
+					// 1. set_pk_j != op_pk_j
+					if is_diff_pk_j {
+						op_pk_x[i][j] = set_pk_j_x;
+						op_pk_y[i][j] = set_pk_j_y;
+					}
```

Comment:
We cant use if statements here, since this would mean that the circuit layout is non deterministic.
We should use SelectChipset when nullifying pks and scores.